_posts/2024-03-22-kraft-migration.md

```diff
@@ -0,0 +1,392 @@
+As a `controller`, it takes part at the quorum controller being a leader or follower, storing the cluster state and handling the metadata changes.
+Effectively, `broker` and `controller` are two different services running on the node within the JVM.
+When in mixed-mode, it gets both roles by taking care of Kafka clients communication on one hand and metadata changes on the other.
+Using the mixed-mode allows to reduce the number of nodes within the cluster compared to a ZooKeeper configuration, even if it is not recommended to be used in a production environment but only for testing or development purposes.
+
+### How to migrate from ZooKeeper to KRaft
+
+As today, it is expected to have users who are running ZooKeeper-based clusters but they are interested in migrating to KRaft mode as soon as possible in order to overcome all the limitations we have talked about.
+Furthermore, ZooKeeper is already considered as deprecated by the Apache Kafka community and its support will be removed with the 4.0 release later this year or early the next year.
+
+There is a manual procedure, made by several phases, to run in order to execute such a migration.
+The following content doesn't want to be an exhaustive description of the migration procedure but more an overview of how it is supposed to work.
+For more details, please refer to the official [ZooKeeper to KRaft Migration](https://kafka.apache.org/documentation/#kraft_zk_migration) documentation.
+
+#### Deploying the KRaft controller quorum
+
+At the beginning, we have the Kafka brokers running in ZooKeeper-mode and connected to the ZooKeeper ensemble used to store metadata.
+
+![ZooKeeper-based cluster](/assets/images/posts/2024-03-22-kraft-migration-01-zk-brokers.png)
+
+> NOTE: the green square boxed number highlights the "generation" of the nodes which are rolled more times during the process.
+
+The first step is about deploying the KRaft controller quorum which will be in charge of storing the metadata in KRaft mode.
+In general, the number of nodes will be the same as the number of the ZooKeeper nodes actually running.
+It is also important to highlight that the migration doesn't support the usage of mixed-nodes.
+The nodes forming the KRaft controller quorum are all configured with the connection to ZooKeeper together with the additional `zookeeper.metadata.migration.enable=true` flag which states the intention to run the migration.
+When the KRaft controllers start, they form a quorum, elect the leader and move in a state where they are waiting for the brokers to register.
+
+![KRaft controller quorum deployed](/assets/images/posts/2024-03-22-kraft-migration-02-kraft-deployed.png)
+
+#### Enabling brokers to run the migration
+
+The next step is about moving the brokers to migration mode.
+In order to do so, the brokers configuration needs to be updated by adding the connection to the KRaft controller quorum and enabling the migration with the `zookeeper.metadata.migration.enable=true` flag.
```

Comment:
`the brokers configuration` -> `the brokers 'configuration`
`adding the connection to the KRaft controller quorum` -> adding connection details for all nodes of the KRaft controller quorum`